peer_control: Make close wait for complete closure, with timeout.

ZmnSCPxj opened the pull request:
Also report tx and txid, and whether we closed unilaterally or
bilaterally, if we could close the channel.

Also make a manpage.

Fixes: #1207
Fixes: #714
Fixes: #622

rustyrussell commented:
This closely mirrors my https://github.com/rustyrussell/lightning/tree/guilt/remove-dev-fail branch which I obviously forgot to put a PR in for!  But yours is better anyway...

I avoided putting in a timeout for simplicity, but OTOH you return more information from close, which is nice but makes the timeout a necessity. You're clearly not from Australia though, because 2 seconds is grossly insufficient as a default :)

You also force a connection initiation to the peer, which is nice.  And wrote the man page!

Can I ask you to cherry-pick commits  or steal stuff as appropriate from my branch?

ZmnSCPxj commented:
> You're clearly not from Australia though, because 2 seconds is grossly insufficient as a default :)

I am also obviously not an embedded semi-independent agent in an AI agent network sharing computational and network resources with other competing and cooperating knowledge agents, especially since 2 seconds would also be grossly insufficient as a default in that situation.

What would be a plausible default for a human being in Australia?

>Can I ask you to cherry-pick commits or steal stuff as appropriate from my branch?

Okay, I shall check.  There are no useful tests for the new closing functionality in particular, so if I see any tests I will steal them.

rustyrussell commented:
30 seconds seems to be the modern default.

ZmnSCPxj commented:
Rebased, changed timeout to 30 seconds, added some code from @rustyrussell , please review again.

cdecker commented on doc/lightning-close.7.txt:
```diff
@@ -0,0 +1,58 @@
+Normally the peer needs to be live and connected in order to negotiate
+a bilateral close.
+Forcing a unilateral close can be used if you suspect you can no longer
+access the peer.
```
access -> contact

cdecker commented on doc/lightning-close.7.txt:
```diff
@@ -0,0 +1,58 @@
+On success, an object with fields 'tx' and 'txid' containing the
+closing transaction are returned.
+It will also have a field 'type' which is either the JSON string
+'bilateral' or the JSON string 'unilateral'.
```
I usually use collaborative close to describe the bilateral close, which is a bit nicer. The spec talks about a "mutual close" or "cooperative close" which is also slightly nicer than bilateral (and you use cooperative in the code later).

ZmnSCPxj replied:
Hmm.  Grepping `lightning-rfc`:

* cooperative: 1 time, in 00-introduction.md
* bilateral: 0 times, ha
* mutual: many times, in 00-introduction.md, 02-peer-protocol.md, 05-onchain.md

Perhaps we should go with 'mutual' instead, as that seems to be the preferred term in the RFC.  Note that the BOLT spec uses 'mutual' and 'unilateral'.

cdecker commented on lightningd/peer_control.c:
```diff
@@ -44,6 +44,17 @@
 #include <wire/gen_onion_wire.h>
 
+struct close_command {
+	/* Off struct lightningd close_commands. */
```
Off -> Of?

cdecker commented on lightningd/peer_control.c:
```diff
@@ -209,13 +220,125 @@ static void remove_sig(struct bitcoin_tx *signed_tx)
+	cc->channel = NULL;
+	command_fail(cc->cmd, "Channel forgotten before proper close.");
+}
+/* Destroy the close command structure. */
```
Nit: one empty line between functions.

cdecker commented on tests/test_lightningd.py:
```diff
@@ -4205,7 +4226,9 @@ def test_peerinfo(self):
 
         # Close the channel to forget the peer
-        l1.rpc.close(l2.info['id'])
+        self.assertRaisesRegex(ValueError,
```
I don't understand why now all of these closes fail in the integration tests. Aren't these supposed to be collaborative closes? The 0 timeout everywhere is a bit excessive IMHO

ZmnSCPxj replied:
The tests assume that `close` does not wait for the close to complete.  This PR makes `close` wait either for the timeout or the close to complete; to emulate the old behavior we set timeout to 0 and let the command fail.  Otherwise I would have to check each instance of `.rpc.close` to see if I can remove the `wait_for_logs` afterwards or if they scan for something more substantial than bilateral close.

cdecker replied:
Well, ok, we should slowly make these closes normal again though (noting this for my cleanup queue).

cdecker commented on lightningd/channel.h:
```diff
@@ -64,6 +64,9 @@ struct channel {
 	struct short_channel_id *scid;
 
+	/* Closing txid, for reporting to JSON */
+	struct bitcoin_txid *closing_txid;
```
This field will not be repopulated when restarting / loading from DB, could that lead us into an inconsistent behavior? Before restart we show the `closing_txid` after the restart we lose that. If we broadcast the transaction we'll have it in the `transactions` table btw.

ZmnSCPxj commented:
Rebased, fixed nits.  Removed `closing_txid` from `listpeers` command instead as I decided it was complicated to have to save it to the DB too, and the new `close` now reports the transaction that actually closes the channel.  Also did some minor cleanups of some calls to the `close` RPC in the test. @cdecker please review again!

cdecker commented:
Thanks @ZmnSCPxj, I oversaw the removal of dev-fail by @rustyrussell, which I think is a bit eager. Forced closes do have the side effect of potentially sending close messages to the peer, which might not be desired in the test cases, so maybe we should keep dev-fail for integration tests?

ZmnSCPxj commented:
Okay, I will remove that commit that removes `dev-fail`.

cdecker commented:
Just kicked Travis, let's see if the failure was a flake

ZmnSCPxj commented:
Well, I notice that the retry system does not seem to work well with this particular test, sigh.

ZmnSCPxj commented:
Can you teach me how to remove the rerunfailures? I tried, to run without rerunfailures, as I suspect that an earlier run will show the root cause, and rerunfailures does not completely clear the test environment, so that reruns will fail for spurious reason once the first run fails for the real reason.

I tried removing the rerunfailures here: https://github.com/ZmnSCPxj/lightning/commit/b8925ff5acbb95c42bc7ee0380fc11e228c589b3 but the CI still runs it with rerun.  Hmm....

ZmnSCPxj commented:
Increasing the timeouts for closing in the `test_closing_different_fees` seems to make it work even in `NO_VALGRIND=0` mode.  Probably just a load problem in the CI (we are running 7 lightningd in parallel and performing 6 closes in parallel).

cdecker commented:
And to answer your question: rerunfailures is enabled automagically on Travis using the following lines in the Makefile:

https://github.com/ElementsProject/lightning/blob/774af5f8176968be895e528586f5569da1daba93/Makefile#L43-L47

So if you set `TRAVIS=false` it'll not rerun.

cdecker commented:
I also missed this comment:

> Removed closing_txid from listpeers command instead as I decided it was complicated to have to save it to the DB too, and the new close now reports the transaction that actually closes the channel.

We are now storing all transactions that we broadcast or that we watched in the `transactions` table, so you could just add the txid into the table and then fetch it from there :wink: But we can add that later as well.

ZmnSCPxj commented:
We already store `last_tx` associated with the channel; the `transactions` table does not have an easy association between channel and transaction.  I just need to figure out when to show it as `closing_tx`/`closing_txid`.  Definitely when it is in `CLOSINGD_COMPLETE`, possibly `ONCHAIN` too (but I am uncertain about the case where we trigger an `error` and the other side gets its commitment onchain before we do; do we track the transaction in this case?).  Let us do it in another PR though, as arguably the important bit (reporting the closing transaction on return of `close` command) is here already and `closing_txid` is an enhancement.
